modules/storage/gdrive.go

```diff
@@ -0,0 +1,497 @@
+		return nil, fmt.Errorf("cannot get root: %v", err)
+	}
+
+	return gdrive, nil
+}
+
+func (gdrive *Gdrive) cacheDelete(path string) error {
+	if cc := cache.GetCache(); cc != nil {
+		return cc.Delete(fmt.Sprintf("gdrive:%s:%s", gdrive.rootDrive.Id, gdrive.fixPath(path)))
+	}
+	return nil
+}
+
+func (gdrive *Gdrive) cacheGet(path string) *drive.File {
+	if cc := cache.GetCache(); cc != nil {
+		if str, ok := cc.Get(fmt.Sprintf("gdrive:%s:%s", gdrive.rootDrive.Id, gdrive.fixPath(path))); ok && str != "" {
+			var node drive.File
+			if err := json.Unmarshal([]byte(str), &node); err != nil {
+				return nil
+			} else if node.Id == "" {
+				return nil
+			}
+			return &node
+		}
+	}
+	return nil
+}
+
+func (gdrive *Gdrive) cachePut(path string, node *drive.File) error {
+	if cc := cache.GetCache(); cc != nil {
+		body, err := json.Marshal(node)
+		if err != nil {
+			return err
+		}
+		return cc.Put(fmt.Sprintf("gdrive:%s:%s", gdrive.rootDrive.Id, gdrive.fixPath(path)), string(body), 60*24)
+	}
+	return nil
+}
+
+func (*Gdrive) URL(path, name string) (*url.URL, error) {
+	return nil, ErrURLNotSupported
+}
+
+// Get Node info and is not trashed/deleted
+func (gdrive *Gdrive) resolveNode(folderID, name string) (*drive.File, error) {
+	name = strings.ReplaceAll(strings.ReplaceAll(name, `\`, `\\`), `'`, `\'`)
+	file, err := gdrive.driveService.Files.List().Fields("*").PageSize(300).Q(fmt.Sprintf(GoogleListQueryWithName, folderID, name)).Do()
+	if err != nil {
+		return nil, err
+	} else if len(file.Files) != 1 {
+		return nil, fs.ErrNotExist
+	} else if len(file.Files) == 0 {
+		return nil, fs.ErrNotExist
+	} else if file.Files[0].Trashed {
+		return file.Files[0], fs.ErrNotExist
+	}
+	return file.Files[0], nil
+}
+
+// List all files in folder
+func (gdrive *Gdrive) listNodes(folderID string) ([]*drive.File, error) {
+	var nodes []*drive.File
+	folderGdrive := gdrive.driveService.Files.List().Fields("*").Q(fmt.Sprintf(GoogleListQuery, folderID)).PageSize(1000)
+	for {
+		res, err := folderGdrive.Do()
+		if err != nil {
+			return nodes, err
+		}
+		nodes = append(nodes, res.Files...)
+		if res.NextPageToken == "" {
+			break
+		}
+		folderGdrive.PageToken(res.NextPageToken)
+	}
+	return nodes, nil
+}
+
+// Get file stream, if error check if is http2 error to make new request
+func (gdrive *Gdrive) GetRequest(node *drive.FilesGetCall) (*http.Response, error) {
+	res, err := node.Download()
+	for i := 0; i < 10 && err != nil; i++ {
+		if urlError, ok := err.(*url.Error); ok {
+			if _, ok := urlError.Err.(http2.GoAwayError); ok || reflect.TypeOf(urlError.Err).String() == "http.http2GoAwayError" {
+				<-time.After(time.Microsecond * 2) // Wait seconds to retry download, to google server close connection
+				res, err = node.Download()
+				continue
+			} else if res != nil && res.StatusCode == 429 {
+				<-time.After(time.Minute) // Wait minutes to reset www.google.com/sorry/index
```

Comment:
If it happens, it is detected as an abuse.

![image](https://github.com/user-attachments/assets/635937c3-fe5c-486c-9389-23f1fd37bd2f)

Reply:
This happens when there are more than 300 requests to the Google API, this happens a lot when there are several people loading an asset or even several seeks are occurring at the same time.

Reply:
Well, that's the problem. I think Google Drive was never designed to work as an Object Storage, it can't be used to store Gitea's files safely.

Reply:
in part as I understand it, I am not [violating terms](https://policies.google.com/terms?hl=en#:~:text=Don%E2%80%99t%20abuse%20our%20services,others%20to%20violate%20these%20terms)

Reply:
But the fact is that the "sorry/index" is reported. Suppose a user is uploading his LFS file or attachment over Gitea to Google Drive, then Google Drive blocks for a while, I do not think the whole system could still run correctly.

Actually, if you'd really like to use Google Drive, there is a simple and nice approach:

https://rclone.org/commands/rclone_serve_s3/

Then Gitea S3 -> rclone serve s3 -> Google Drive